RecyclerView inside ViewPager

chiragagrawal-tudip opened the pull request:
- Used recyclerview inside viewpager and viewpager is inside the slider. But the issue we are getting from this library is that when we scroll up recyclerview, slider gets stuck in between.
- So we have done some changes to solve it in SlidingLayer file.
- I have tested my changes with your library and it's working fine. That is not disturbed by my code.
- With that, my issue has also been resolved.

chiragagrawal-tudip committed:
- Used recyclerview inside viewpager and viewpager is inside the slider. But the issue we are getting from this library is that when we scroll up recyclerview, slider gets stuck in between.
- So we have done some changes to solve it in SlidingLayer file.
- I have tested my changes with your library and it's working fine. That is not disturbed by my code.
- With that, my issue has also been resolved.

JlUgia commented on Library/src/main/java/com/wunderlist/slidinglayer/SlidingLayer.java:
```diff
@@ -38,6 +38,7 @@
 import android.support.v4.view.ViewCompat;
 import android.support.v4.view.ViewConfigurationCompat;
 import android.util.AttributeSet;
+import android.util.Log;
```
This was probably forgotten in between. Can be taken out.

JlUgia commented on Library/src/main/java/com/wunderlist/slidinglayer/SlidingLayer.java:
```diff
@@ -864,10 +865,14 @@ protected boolean canScroll(View v, boolean checkV, int dx, int dy, int x, int y
+        boolean temp = checkV && (
                 (allowedDirection() == HORIZONTAL && ViewCompat.canScrollHorizontally(v, -dx) ||
-                        allowedDirection() == VERTICAL && ViewCompat.canScrollVertically(v, -dy)));
+                        allowedDirection() == VERTICAL && v.canScrollVertically( -dy)));
```
There should be a benefit on using the ViewCompat library. That will take away some of the headaches of dealing with the two different API libraries. Unless this is deprecated based on the minSdkVers. Is that the case?

chiragagrawal-tudip replied:
ViewCompat is deprecated in our project because our minSdkVers. is 18.

JlUgia replied:
I see. And doesn't the `ViewCompat` scroll check work properly? The minimum sdk supported is something that there is no motivation to change at the moment, hence this change is something that will presumably have to stay at your fork.

JlUgia commented on Library/src/main/java/com/wunderlist/slidinglayer/SlidingLayer.java:
```diff
@@ -864,10 +865,14 @@ protected boolean canScroll(View v, boolean checkV, int dx, int dy, int x, int y
                 (allowedDirection() == HORIZONTAL && ViewCompat.canScrollHorizontally(v, -dx) ||
-                        allowedDirection() == VERTICAL && ViewCompat.canScrollVertically(v, -dy)));
+                        allowedDirection() == VERTICAL && v.canScrollVertically( -dy)));
+        if (v instanceof RecyclerView && dy > 0) {
```
I'd suggest we turn this into some form a flag passed in from the top of the lib, where the view is added, so that we can avoid adding a complete dependency just to do a class check. Any other ideas?

chiragagrawal-tudip replied:
Can you please elaborate more on this?

JlUgia replied:
I was wondering whether there was a way to obtain the same outcome without needing to let the lib know about `RecyclerView`. There are two paths that come to my mind:
1. Pass a flag into the configuration of the library (eg.: `layer.setVerticallyScrollingChildren(true)`), and use that instead of the `instanceof` check.
2. Adding the support library to dependencies and use interfaces like [`ScrollingView`](https://developer.android.com/reference/android/support/v4/view/ScrollingView) to make this work not only with `RecyclerView` but also with other implementers of the interface.

Now, mentioned all this, I'm reading at the code under `RecyclerView` and `canScrollVertically` should be sufficient to determine whether the children can scroll vertically or not. What's your specific use case, and why that does not work by itself?

chiragagrawal-tudip replied:
Please see this video when we scroll down recycler view, Slider view gets stuck https://www.youtube.com/watch?v=4oByYVkV-dM

And if recycler view is not scrolled down slider works with no issues: https://www.youtube.com/watch?v=cpwMz34UKwE

JlUgia replied:
Interesting. It looks like a different problem to me. These buttons don't seem to be part of the RecyclerView are they? Are these custom buttons? Regular ones should not capture de click event if you are dragging it. Or is the button header part of the recyclerView too?

chiragagrawal-tudip replied:
Those buttons are custom tab views. But they are not inside the recyclerview.

JlUgia replied:
Ok, that's most likely the reason. What you can do to confirm this behavior is to use a built-in clickable component such as `Button` or `ImageButton`, and verify whether the sliding layer responds or not.
Once confirmed, you just need to get the delivery of touch events right inside of your components, by overriding `onInterceptTouchEvent` and only capture the event if the interaction is part of a press and not a drag initiative. You can read more about it in [here](https://developer.android.com/training/gestures/viewgroup#intercept).

Said so, if your custom component is inheriting from a built in clickable component, that should have been handled already there. What type of custom component is your tab? Happy to take a loot to the source if that could help in any way.

chiragagrawal-tudip replied:
Ok but I don't think issue is of tab bar because if I don't scroll recycler view down, it works perfectly fine. and if I scroll down a little it stops responding.

JlUgia replied:
What happens if you try to pull the sliding layer down from the recycler view? It might be a bit tricky to troubleshoot this using this comment window only. I'm happy to take a look at the code or jump on a video to talk about it. 

I think that trying to replace the tab with a regular button is a very quick check that could shed some light into this.
Let me know.

chiragagrawal-tudip replied:
can we have skype call, I can share my screen, let me know your skypeID.
Thanks

JlUgia replied:
Let's use a Hangout so that we avoid sharing private info in here. What time works best with you? I can meet all throughout the day today.

chiragagrawal-tudip replied:
Please let me know, Can we meet today?

JlUgia replied:
Can do in the next half an hour? Otherwise, would something like 19:30 work?

chiragagrawal-tudip replied:
Ok Sure, My hangout ID is chirag.agrawal@tudip.nl.
Please ping me, whenever you are free.

JlUgia replied:
Ok, looks like this is not necessary anymore.

chiragagrawal-tudip replied:
I have raised another PR (https://github.com/wunderlist/android-sliding-layer-lib/pull/112).
Please have a look on it.

chiragagrawal-tudip replied:
Hi @JlUgia,
Please have a look at PR (https://github.com/wunderlist/android-sliding-layer-lib/pull/112). 
This will not add any extra dependency.

chiragagrawal-tudip commented:
- Removed previous changes.
- I have implemented a method 'canScrollToView()' in which I have just added the return expression of 'canScroll()' which is holding the original return boolean value.
- It will not affect the library at all.
- I am overriding the 'canScrollToView()' method in the custom slider that we are using in the project as required by using the 'view' and 'dy' of method 'canScroll()'.
- I have raised another PR (https://github.com/wunderlist/android-sliding-layer-lib/pull/112) and closing this PR.

chiragagrawal-tudip commented:
Hi @JlUgia,
Please have a look at PR (https://github.com/wunderlist/android-sliding-layer-lib/pull/112).
